### PR TITLE
[CPU]GatherElements support dynamic shape

### PR DESCRIPTION
**Pull request by luo-cheng2021**
### Details:
 - *Added dynamism support for GatherElements*
 - *Updated CPU tests to cover dynamic shapes*

### Tickets:
 - *70994*

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_gather_elements_node.cpp [hunk 1]
```diff
@@ -45,29 +40,40 @@ MKLDNNGatherElementsNode::MKLDNNGatherElementsNode(const std::shared_ptr<ngraph:
     if (op->get_input_size() != 2 || op->get_output_size() != 1)
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_gather_elements_node.cpp:40]:
```suggestion
    if (inputShapes.size() != 2 || outputShapes.size() != 1)
```

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_gather_elements_node.cpp [hunk 2]
```diff
@@ -45,29 +40,40 @@ MKLDNNGatherElementsNode::MKLDNNGatherElementsNode(const std::shared_ptr<ngraph:
...
         IE_THROW() << errorPrefix_ << " has invalid number of input/output edges.";
 
-    const auto& dataDims = op->get_input_shape(dataIndex_);
-    const auto& indicesDims = op->get_input_shape(indicesIndex_);
-    if (dataDims.size() != indicesDims.size())
+    const auto dataRank = inputShapes[dataIndex_].getRank();
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_gather_elements_node.cpp:43]: use `getInputShapeAtPort/getOutputShapeAtPort`

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_gather_elements_node.cpp [hunk 3]
```diff
@@ -45,29 +40,40 @@ MKLDNNGatherElementsNode::MKLDNNGatherElementsNode(const std::shared_ptr<ngraph:
...
+    if (inputShapesDefined()) {
+        if (needPrepareParams())
+            prepareParams();
+        updateLastInputDims();
+    }
+}
 
-    auto outputShape = op->get_output_shape(0);
+void MKLDNNGatherElementsNode::prepareParams() {
+    const auto& dataDims = getParentEdgeAt(dataIndex_)->getMemory().getStaticDims();
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_gather_elements_node.cpp:66]:
```suggestion
    const auto& dataDims = getParentEdgesAtPort(dataIndex_)[0]->getMemory().getStaticDims();
```

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/gather_elements.cpp [hunk 4]
```diff
@@ -2,88 +2,142 @@
...
 
-        auto params = ngraph::builder::makeParams(ngDPrc, {dataShape});
-        auto activation = ngraph::builder::makeGatherElements(params[0], indicesShape, ngIPrc, axis);
-        activation->get_rt_info() = getCPUInfo();
-        function = std::make_shared<ngraph::Function>(ngraph::NodeVector{activation}, params, "GatherElements");
+        auto gather = std::make_shared<ngraph::op::v6::GatherElements>(
+            params[0], params[1], axis);
+        gather->get_rt_info() = getCPUInfo();
+        ngraph::ResultVector results{std::make_shared<ngraph::opset1::Result>(gather)};
+        function = std::make_shared<ngraph::Function>(results, params, "GatherElements");
```
Comment (mandrono) [inline inference-engine/tests/functional/plugin/cpu/single_layer_tests/gather_elements.cpp:110]:
```suggestion
        function = makeNgraphFunction(dPrecision,  params, gather,"GatherElements");
```

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/gather_elements.cpp [hunk 5]
```diff
@@ -2,88 +2,142 @@
...
 }
 
-
 namespace {
 std::vector<CPUSpecificParams> cpuParams_4D = {
         CPUSpecificParams({nchw}, {nchw}, {}, {})
 };
 
+const std::vector<std::vector<InputShape>> inDynamicShapeParams = {
+    {{{ngraph::Dimension::dynamic(), ngraph::Dimension::dynamic(), ngraph::Dimension::dynamic(), ngraph::Dimension::dynamic()},
```
Comment (mandrono) [inline inference-engine/tests/functional/plugin/cpu/single_layer_tests/gather_elements.cpp:124]: you can use `-1` instead `ngraph::Dimension::dynamic()`
and `{1, 10}` instead `ngraph::Dimension(1, 10)`

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_gather_elements_node.cpp (hunk 2 above)
Comment (luo-cheng2021) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_gather_elements_node.cpp]: Done

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/gather_elements.cpp (hunk 4 above)
Comment (luo-cheng2021) [reply inference-engine/tests/functional/plugin/cpu/single_layer_tests/gather_elements.cpp]: Done.

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/gather_elements.cpp (hunk 5 above)
Comment (luo-cheng2021) [reply inference-engine/tests/functional/plugin/cpu/single_layer_tests/gather_elements.cpp]: Done.

**Review by mandrono — APPROVED** [review]

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**